File: timus/1291.pas
var
	n, i, k : longint;
	a : array [1..1000] of longint;

	g : array [1..1000] of longint;
	next, num : array [1..10000] of longint;
	last : longint;

	first, vel : longint;

	nom, den : array [1..1000] of longint;
	v : array [1..1000] of boolean;

function GCD(a, b : longint) : longint;	
	var
		c : longint;

	begin
		while (b <> 0) do
		begin
			c := a mod b;
			a := b;
			b := c;
		end;
		GCD := a;
	end;

procedure DFS(i : longint);
	var
		j, l, gg : longint;

	begin
		v[i] := true;
		j := g[i];
		while (j <> 0) do
		begin
			l := num[j];
			if (not v[l]) then
			begin
				nom[l] := -nom[i]*a[i];
				den[l] := den[i]*a[l];
				gg := GCD(Abs(nom[l]), Abs(den[l]));

				nom[l] := nom[l] div gg;
				den[l] := den[l] div gg;

				DFS(l);
			end;

			j := next[j];
		end;
	end;

begin
	last := 0;
	Fillchar(g, sizeof(g), 0);

	Read(n);
	for i := 1 to n do
	begin
		Read(a[i]);
		Read(k);
		while (k <> 0) do
		begin
			Inc(last);
			next[last] := g[i];
			num[last] := k;
			g[i] := last;

			Inc(last);
			next[last] := g[k];
			num[last] := i;
			g[k] := last;

			Read(k);
		end;
	end;

	Read(first, vel);

	for i := 1 to n do
	begin
		nom[i] := 0;
		den[i] := 1;
	end;

	nom[first] := vel;

	Fillchar(v, sizeof(v), 0);
	DFS(first);

	for i := 1 to n do
		Writeln(nom[i], '/', den[i]);
end.

File: timus/1291.c
#include <stdio.h>
#include <string.h>

int n, c[1000], b[1000], g[1000], p[1000], q[1000], ln[2000], lv[2000];

int gcd(int a, int b) {
	int c;

	if (a < 0)
		a = -a;

	while (b != 0) {
		c = a % b;
		a = b;
		b = c;
	}

	return a;
}

void dfs(int u) {
	int i, v, d, pp, qq;

	b[u] = 1;

	i = g[u];
	while (i != -1) {
		v = lv[i];
		if (!b[v]) {
			pp = -p[u] * c[u]; qq = q[u] * c[v];
			d = gcd(pp, qq); pp /= d; qq /= d;
			p[v] = pp; q[v] = qq;

			dfs(v);
		}
		i = ln[i];
	}
}

int main() {
	int u, v, i, s, vv, lf;

	scanf("%d", &n);
	for (i = 0; i < 2*n; i++)
		ln[i] = i + 1;
	lf = 0;
	for (u = 0; u < n; u++) {
		g[u] = -1;
		scanf("%d%d", &c[u], &v);
		while (v != 0) {
			v--;

			i = lf; lf = ln[lf];
			ln[i] = g[u]; g[u] = i;
			lv[i] = v;

			scanf("%d", &v);
		}
	}

	scanf("%d%d", &s, &vv); s--;

	for (u = 0; u < n; u++) {
		p[u] = 0;
		q[u] = 1;
	}
	p[s] = vv;
	
	memset(b, 0, sizeof(b));
	dfs(s);

	for (u = 0; u < n; u++)
		printf("%d/%d\n", p[u], q[u]);

	return 0;
}
